src/algebra/monoid_algebra/degree.lean

```diff
@@ -0,0 +1,191 @@
+/-
+Copyright (c) 2022 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import algebra.monoid_algebra.basic
+
+/-!  #  Max-degree and min-degree of an `add_monoid_algebra`
+
+Let `R` be a semiring and let `A` be a `semilattice_sup`.
+
+For an element `f : add_monoid_algebra R A`, this file defines
+* `add_monoid_algebra.max_degree`: the max-degree taking values in `with_bot A`,
+* `add_monoid_algebra.min_degree`: the min-degree taking values in `with_top A`.
+If the grading type `A` is a linearly ordered additive monoid, then these two notions of degree
+coincide with the standard one:
+* the max-degree is the maximum of the exponents of the monomials that appear with non-zero
+  coefficient in `f`, or `⊥`, if `f = 0`;
+* the min-degree is the minimum of the exponents of the monomials that appear with non-zero
+  coefficient in `f`, or `⊤`, if `f = 0`.
+
+The main results are
+* `max_degree_mul_le` -- the max-degree of a product is at most the sum of the max-degrees,
+* `le_min_degree_mul` -- the min-degree of a product is at least the sum of the min-degrees,
+* `max_degree_add_le` -- the max-degree of a sum is at most the sup of the max-degrees,
+* `le_min_degree_add` -- the min-degree of a sum is at least the inf of the min-degrees.
+
+## Implementation notes
+
+The current plan is to state and prove lemmas about `finset.support_sup` with a "generic"
+degree/weight function `D` from the grading Type `A` to a somewhat ordered Type `B`.
+
+Next, the general lemmas get specialized twice:
+* once for `max_degree` (essentially a simple application) and
+* once for `min_degree` (a simple application, via `order_dual`).
+These final lemmas are the ones that likely get used the most.  The generic lemmas about
+`finset.support.sup` may not be used directly much outside of this file.
+
+To see this in action, you can look at the triple
+`(sup_support_mul_le, max_degree_mul_le, le_min_degree_mul)`.
+-/
+
+variables {R A B ι : Type*} [semilattice_sup B] [order_bot B]
+
+namespace add_monoid_algebra
+open_locale classical big_operators
+
+section general_results_assuming_semilattice_sup
+
+section semiring
+
+variables [semiring R]
+
+section add_only
+
+lemma sup_support_add_le (D : A → B) (f g : add_monoid_algebra R A) :
+  (f + g).support.sup D ≤ (f.support.sup D) ⊔ (g.support.sup D) :=
+begin
```

Comment:
Proved with `order_dual`s.  Some of the proofs could be shorter, once the first statement is around, but I went for mechanical dualization, instead of reproducing the older proofs.